haystack/nodes/other/route_documents.py

```diff
@@ -38,43 +51,89 @@ def __init__(self, split_by: str = "content_type", metadata_values: Optional[Lis
+        for idx, item in enumerate(metadata_values):
+            if isinstance(item, list):
+                if value in item:
+                    return idx
+            else:
+                if value == item:
+                    return idx
+        return len(metadata_values)
+
+    def _split_by_metadata_values(
+        self, metadata_values: Union[List, List[List]], documents: List[Document]
+    ) -> Dict[str, List[Document]]:
+        split_documents: Dict[str, List] = {f"output_{i + 1}": [] for i in range(len(metadata_values) + 2)}
+        remaining_key = f"output_{len(metadata_values) + 1}"
+
+        for doc in documents:
+            current_metadata_value = doc.meta.get(self.split_by, remaining_key)
+            index = self._get_metadata_values_index(metadata_values, current_metadata_value)
+            split_documents[f"output_{index + 1}"].append(doc)
+
+        if not self.return_remaining:
+            if logger.isEnabledFor(logging.WARNING) and len(split_documents[remaining_key]) > 0:
+                logger.warning(
+                    "%s documents were skipped because they were either missing the metadata field '%s' or the"
+                    " corresponding metadata value is not included in `metadata_values`.",
+                    len(split_documents[remaining_key]),
+                    self.split_by,
+                )
+            del split_documents[remaining_key]
+
+        return split_documents
 
     def run(self, documents: List[Document]) -> Tuple[Dict, str]:  # type: ignore
         if self.split_by == "content_type":
-            split_documents: Dict[str, List[Document]] = {"output_1": [], "output_2": []}
-
-            for doc in documents:
-                if doc.content_type == "text":
-                    split_documents["output_1"].append(doc)
-                elif doc.content_type == "table":
-                    split_documents["output_2"].append(doc)
-
+            split_documents = self._split_by_content_type(documents)
         else:
-            assert isinstance(self.metadata_values, list), (
-                "You need to provide metadata_values if you want to split" " a list of Documents by a metadata field."
-            )
-            split_documents = {f"output_{i+1}": [] for i in range(len(self.metadata_values))}
-            for doc in documents:
-                current_metadata_value = doc.meta.get(self.split_by, None)
-                # Disregard current document if it does not contain the provided metadata field
-                if current_metadata_value is not None:
-                    try:
-                        index = self.metadata_values.index(current_metadata_value)
-                    except ValueError:
-                        # Disregard current document if current_metadata_value is not in the provided metadata_values
-                        continue
-
-                    split_documents[f"output_{index+1}"].append(doc)
-
+            assert isinstance(
```

Comment:
I originally wanted to leave this assert as a `raise ValueError` in the init method, but mypy complained that `self.metadata_values` could be `None` at this point in the code which is why I added the assert back.

Reply:
@silvanocerza do you think it's worth keeping this assert or should I add an ignore statement for mypy?

Reply:
I prefer not to ignore statements to be fair. `assert` is good enough, explicit check and throw would be even better I think.

Reply:
Okay sounds good. The explicit check and throw still caused mypy to throw an error, which is why I went with the assert.

Reply:
That's strange, I get no errors if I change the `else:` to `elif self.metadata_values:`. 🤔

Reply:
Ahh okay I didn't do that it way before, I'll go ahead and change it.

Reply:
done!